### ISSUE TITLE
Export our variable and prepend relative directories to @INC

### ISSUE THREAD

**Issue by XSven** [hkoba/p5-File-AddInc#2]
This is a feature request refering to the 0.001 version of `File::AddInc`.

Currently `File::AddInc` prepends `$libdir` created with the class method `libdir()` to `@INC`.

1. `use File::AddInc qw($foo);` should create an `our` variable with the name `$foo` and the value `$libdir`. The value is of course  still prepended to `@INC`.
2. `use File::AddInc qw(bar);` should prepend `$libdir` and `$libdir/bar` to `@INC`. `$libdir` should be the 0th element of `@INC` and `$libdir/bar` should be the 1st element of `@INC`.
3. `use File::AddInc qw($foo bar);` should create an `our` variable with the name `$foo` and the value `$libdir` and should prepend `$libdir` and `$libdir/bar` to `@INC`.
4. `use File::AddInc qw(baz $foo bar);` should create an `our` variable with the name `$foo` and the value `$libdir` and should prepend `$libdir` and `$libdir/baz` and `$libdir/bar` to `@INC`.

**Comment by hkoba** [comment]
Hi!
Thank you for your interesting suggestions. 

Your request item 1( `use File::AddInc qw($foo);`) sounds enough general (except the modification of `@INC`) and nice to have, but other items 2〜4 sounds quite odd to me. 

So, before changing the code, I want to hear more about your use cases.
Assume we had the item 1( `use File::AddInc qw($foo);`), can you achieve your goals in conjunction with `use lib`?
In other words, can following codes satisfy your need?

```perl
# For item 2. use File::AddInc qw(bar); 
use File::AddInc qw($libdir);
use lib $libdir, "$libdir/bar";

# For item 3. use File::AddInc qw($foo bar);
use File::AddInc qw($foo);
use lib $foo, "$foo/bar";

# For item 4. use File::AddInc qw(baz $foo bar);
use File::AddInc qw($foo);
use lib "$foo/baz", $foo, "$foo/bar";
```

**Comment by XSven** [comment]
The `use lib` statement that refers to item 4. should be

`use lib $foo, "$foo/baz",  "$foo/bar";`

The order matters. But in the end you are right. I wouldn't concatenate the path elements together. I would use File::Spec->catfile. Furthermore a `-d` file test is needed because only existing directories should be prepended to `@INC`. Last but not least we are writing all our applications as modulinos and the extra `use lib` line would add redundant boilerplate code.

A concrete use case at our end would be

`use File::AddInc qw(etc perl5);`

`"$libdir/etc"` would contains module like configuration files.
`"$libdir/perl5"` would contain additional CPAN modules installed with `local::lib`.

**Comment by XSven** [comment]
I have thought about the usecase number 4 again, and I think `use File::AddInc qw(baz $foo bar);` should really be interpreted as an instruction to `use lib "$foo/baz", $foo, "$foo/bar";` in the same order as defined by the `File::AddInc` import list. This offers complete flexibility and transparency to the user.

**Comment by hkoba** [comment]
Hi!

I pushed the change to use `use File::AddInc qw($libdir);` to branch `tkt/2-libdir_var`. You can see the test code for it [here](https://github.com/hkoba/p5-File-AddInc/blob/tkt/2-libdir_var/t/gh/2-libdir_var/MyApp.pm#L7). I hope this can satisfy your need.

Also, I added another form of use of File::AddInc. That is:

```perl
use File::AddInc [these_libdirs => 'etc', q{}, 'perl5'];
```

Roughly, above is an equivalent of the following: 
```perl
use File::AddInc qw($foo);
use lib "$foo/etc", $foo, "$foo/perl";
```

If you have any suggestions about this feature (especially about the name of the option `these_libdirs`), please let me know.

**Comment by XSven** [comment]
I am astonished about the complexity that you have added:
- You have added named parameters `libdir_var` and `these_libdir`, although the dollar (`$`) sigil would have been enough to find the variable name that should be exported as a global `our` variable.
- Although TIMTOWTDI is a perl mantra, I don't like it, it increases code complexity and it reduces the maintanability for others.

- I do not see the need to offer

`use File::AddInc [libdir_var => qw($libdir)];`

if you have offered that

`use File::AddInc qw($libdir);`

With this approach you have also introduced arrayref import arguments:-(
- Finally

`use File::AddInc [these_libdirs => 'etc', q{}, 'perl5'];`
offers room for misinterpretation. From what you have written, the `these_libdirs` name refers to all the following 3 values. Usually if am using the arrow form instead of the comma form to separate values, the arrow form refers to the next value only. I would have expected something like this

`use File::AddInc these_libdirs => ['etc', q{}, 'perl5'];`

Please do not implement this! Think about an approach again that supports a straight and easy interface that I have originally described in my feature request (of course with item 4. fixed with respect to the ordering topic).

**Comment by hkoba** [comment]
I can understand you feel odd to the array form parameters of "use" in my previous example. But I don't withdraw it **(at least internally)** from my code. I have a strong reason to do it. That form is my answer to solve the problem of the extensibility of exporter functionalities.

## Why `[pragma => @args]`

I have been long frustrated with that duplicate code in many exporters (mine and existing ones from others). Most of their implementations are directly written in `import()` function with hard-coded dispatch chain with `if $keyword elsif $keyword2...`. It requires complex, un-intuitive code to extend specific keyword behavior through subclassing of the exporter. (Which usually scans `@_`, add/remove specific keywords, do something and call original with `goto $pack->can("import")` ...).

To make things worse, we write exporters to treat its arguments entirely free as we like. Some exporter takes imported symbols (like "vars.pm"), and others accept ":SOME_SPECIAL_KEYWORD" sequence (like "open.pm"). This freedom costs on us. We must always read into the implementation of "import()" to extend exporters written by others.

If I understand it correctly, this extensibility problem is caused by the lack of separation between "hook" and "interface of actual functionalities". "import()" functions of each package are "hook" (entry-point function) to catch "use" and get "caller()". Unfortunately, most exporters do not have public interfaces to call their core functionalities without involving "caller()". They only have "import()", so users are forced to either struggle with them or steal the code and modify it.

That's why I introduced `[pragma => @args]` form arguments to my exporters. In my exporters, "import()" only do limited tasks. Save the "caller()" and scan the arguments to dispatch specified tasks. I call such behavior as **"Meta-Object Protocol for Import"** (or MOP4Import for short).

For example, my "MOP4Import::Declare" treats each argument as a method call named starting with "declare_".  `-as_base` invokes `$pack->declare_as_base($caller)`. `[base => qw/mod1 mod2/]` invokes `$pack->declare_base($caller, qw/mod1 mod2)`.  There are also other style of arguments treatment (ex. `name => $pragmaArray, name2 => $pragmaArray, ...`), but anyway they just map given arguments to method calls in specific manners, uniformly.

Finally, if I implement exporters in this way, you can freely extend it through subclassing with your own "import()" function with your own flavor of argument treatment. You can just capture caller and invoke `$pack->declare_these_libdirs($caller_opts, @_)`.

### About code complexity and maintainability

You mentioned about the increased complexity of my code and worried about its maintainability. But what you should compare with my code is other exporters. If you know better exporter architecture/implementation with respect to extensibility and maintainability than mine, please let me know.

**Comment by XSven** [comment]
Ok, it seems you have gained much more (negative) experience with existing `import()` method implementation than me.  I have added  a subtest that both uses the `libdir_var` and the `these_libdirs` key

```
subtest q{use File::AddInc [libdir_var => '$libdir', these_libdirs => 'etc', 'perl5']}, sub {
  my $testDir = File::Spec->rel2abs($testName);
  my $targetFile = "MyApp5.pm";

  is qx($^X -I$distDir/lib $testDir/$targetFile), "FOObar\n", "these_libdirs works";
};
```
to `t/gh/2-libdir_var/test.t`. The `MyApp5.pm` looks like this

```
#!/usr/bin/env perl
package
  MyApp5;
use strict;
use warnings;

use File::AddInc [libdir_var => '$libdir', these_libdirs => 'etc', 'perl5'];

use my_config;
use my_lib;

unless (caller) {
  print $my_config::foo, my_lib::bar(), "\n";
}

1;
```
The test fails with 

Can't locate my_config.pm in @INC (...)

**Comment by hkoba** [comment]
Please change this line:

```perl
use File::AddInc [libdir_var => '$libdir', these_libdirs => 'etc', 'perl5'];
```

to the following:

```perl
use File::AddInc 
  [libdir_var => '$libdir'], 
  [these_libdirs => 'etc', 'perl5'],
;
```

Not tested, but expected to work.

**Comment by XSven** [comment]
Thanks it works.

**Comment by XSven** [comment]
When are you going to create a released CPAN version?

**Comment by hkoba** [comment]
Probably within this week.

By the way, do you want to include above `MyApp5.pm` into coming release?
If so, please consider to submit ~the~ a pull request for it.
